### PR TITLE
move ShortName to bevy_reflect

### LINKED ISSUE #15339: Move ShortName to bevy_reflect
## What problem does this solve or what need does it fill?

A step toward minimizing bevy_utils in accordance with #11478

### PR DESCRIPTION
**Pull request by BenjaminBrienen**
# Objective

- Goal is to minimize bevy_utils #11478

## Solution

- Move the file short_name wholesale into bevy_reflect

## Testing

- Unit tests
- CI

## Migration Guide

- References to `bevy_utils::ShortName` should instead now be `bevy_reflect::ShortName`.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
Oh good, I was worried this would mess up the feature flags for `bevy_ecs`. Nice to see this work cleanly.

I think this makes sense there: it's a tool to work with type names.

**Comment by alice-i-cecile** [comment]
Can you update your migration guide to be a bit more descriptive? It doesn't actually describe the fix when read outside of the context of this PR.

File: crates/bevy_hierarchy/src/valid_parent_check_plugin.rs [hunk 1]
```diff
@@ -3,8 +3,9 @@ use std::marker::PhantomData;
 #[cfg(feature = "bevy_app")]
 use crate::Parent;
 use bevy_ecs::prelude::*;
+use bevy_reflect::ShortName;
```
Comment (mockersf) [inline crates/bevy_hierarchy/src/valid_parent_check_plugin.rs:6]:
```suggestion
#[cfg(feature = "bevy_app")]
use bevy_reflect::ShortName;
```
Shouldn't this still be behind the feature?

File: crates/bevy_reflect/Cargo.toml [hunk 2]
```diff
@@ -25,6 +25,7 @@ debug_stack = []
 documentation = ["bevy_reflect_derive/documentation"]
 # Enables function reflection
 functions = ["bevy_reflect_derive/functions"]
+alloc = []
```
Comment (mockersf) [inline crates/bevy_reflect/Cargo.toml:28]: to keep everything as it was, this should be part of the default feature of the bevy_reflect crate

Comment (BenjaminBrienen) [reply crates/bevy_reflect/Cargo.toml:28]: done

File: crates/bevy_reflect/src/lib.rs [hunk 3]
```diff
@@ -549,6 +549,8 @@ mod reflect;
 mod reflectable;
 mod remote;
 mod set;
+mod short_names;
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/lib.rs:552]: Minor nit: why not just "short_name" (i.e. singular)? I feel like that reads a little better: `short_name::ShortName`.

File: crates/bevy_reflect/src/lib.rs [hunk 4]
```diff
@@ -549,6 +549,8 @@ mod reflect;
...
+pub use short_names::ShortName;
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/lib.rs:553]: Nit: Can we move this export so it's with the other exports (and yes, the mods and exports are a bit all over the place in this file haha)?

**Review by MrGVSV — APPROVED** [review]

File: crates/bevy_reflect/src/lib.rs (hunk 3 above)
Comment (MrGVSV) [reply crates/bevy_reflect/src/lib.rs:552]: Ah, looks like that was the original module name. I think we can rename it here.

**Review by james7132 — APPROVED** [review]

**Comment by hymm** [comment]
> Oh good, I was worried this would mess up the feature flags for bevy_ecs. Nice to see this work cleanly.

Given that `use_shortnames` for schedules needed to be feature flagged behind bevy reflect, I'm not sure that this is a clean change. That feature is about using shortnames when reporting ambiguities and has nothing to do with bevy reflect.

**Comment by BenjaminBrienen** [comment]
> > Oh good, I was worried this would mess up the feature flags for bevy_ecs. Nice to see this work cleanly.
> 
> Given that `use_shortnames` for schedules needed to be feature flagged behind bevy reflect, I'm not sure that this is a clean change. That feature is about using shortnames when reporting ambiguities and has nothing to do with bevy reflect.

I think I see what you mean now. What do you suggest as a solution, keeping in mind we, ideally, don't want a `utils` crate.

**Comment by hymm** [comment]
I think the only real options are to move it into it's own crate in the bevy repo or move it out like we did with `atomicow`.

**Pull request merged**